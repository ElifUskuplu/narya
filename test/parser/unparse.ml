open Testutil.Repl

let () =
  run @@ fun () ->
  assume "A" "Type";
  def "nat" "Type" "(A → A) → (A → A)";
  def "zero" "nat" "f x ↦ x";
  def "one" "nat" "f x ↦ f x";
  print "one";
  def "two" "nat" "f x ↦ f (f x)";
  def "three" "nat" "f x ↦ f (f (f x))";
  def "plus" "nat → nat → nat" "m n f x ↦ (m f) (n f x)";
  print "plus one one";
  print "plus two three";
  def "times" "nat → nat → nat" "m n f x ↦ m (n f) x";
  print "times one one";
  print "times two three";
  print "nat";
  print "(X:Type) → X → X";
  print "(X:Type)(_: X) → X";
  print "(x:A)(y:A) → A → (z:A) → A";
  def "unat" "Type" "(X:Type) → (X→X) → X→X";
  def "uzero" "unat" "_ _ x ↦ x";
  def "uone" "unat" "_ f x ↦ f x";
  def "utwo" "unat" "_ f x ↦ f (f x)";
  def "uthree" "unat" "_ f x ↦ f (f (f x))";
  print "uzero";
  print "uone";
  def "exp" "unat → unat → unat" "m n Y f x ↦ n (Y→Y) (m Y) f x";
  print "exp";
  print "exp utwo uthree"
