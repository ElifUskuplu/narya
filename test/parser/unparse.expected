f x ↦ f x
f x ↦ f (f x)
f x ↦ f (f (f (f (f x))))
f x ↦ f x
f x ↦ f (f (f (f (f (f x)))))
(A → A) → A → A
(X : Type) → X → X
(X : Type) → X → X
(x : A) (y : A) → A → (z : A) → A
_ _ x ↦ x
_ f x ↦ f x
m n Y f x ↦ n (Y → Y) (x0 x1 ↦ m Y (x2 ↦ x0 x2) x1) (x0 ↦ f x0) x
Y f x ↦ f (f (f (f (f (f (f (f x)))))))
{ fst ≔ zero. ; snd ≔ zero. }
{ fst ≔ x ↦ x ; snd ≔ suc. (suc. zero.) }
s .fst (suc. (suc. (suc. zero.)))
zero.
refl b
a ↦ refl a
b^00
b^000
b
